Define CNB_BUILD_CONFIG_DIR behavior (#345)
buildpacks/spec: platform.md

  [-app <app>] \
  [-analyzed <analyzed>] \
  [-buildpacks <buildpacks>] \

>   [-build-config <buildConfig>]

Reviewer: ```suggestion
  [-build-config <build-config>]
```

| `<log-level>`   | `CNB_LOG_LEVEL`        | `info`                                                 | Log Level                                                                                                                                                    |
| `<order>`       | `CNB_ORDER_PATH`       | `<layers>/order.toml` if present, or `/cnb/order.toml` | Path resolution for order definition (see [`order.toml`](#ordertoml-toml))                                                                                   |
| `<plan>`        | `CNB_PLAN_PATH`        | `<layers>/plan.toml`                                   | Path to output resolved build plan                                                                                                                           |
| `<platform>`    | `CNB_PLATFORM_DIR`     | `/platform`                                            | Path to platform directory                                                                                                                                   |

> | `<buildConfig>` | `CNB_BUILD_CONFIG_DIR` | `/cnb/config/env.build`                                | Path to operator defined environment variables                                                                                                               |

Reviewer 2: ```suggestion
| `<build-config>` | `CNB_BUILD_CONFIG_DIR` | `/cnb/config/env.build`                                | Path to operator defined environment variables                                                                                                               |
```

> | `<buildConfig>` | `CNB_BUILD_CONFIG_DIR` | `/cnb/config/env.build`                                | Path to operator defined environment variables                                                                                                               |

Reviewer 2: ```suggestion
| `<buildConfig>` | `CNB_BUILD_CONFIG_DIR` | `/cnb/build-config`                                | Path to operator-defined build config                                                                                                               |
```

Reviewer 2: This matches the implementation in the lifecycle... also if I'm not mistaken the build config dir is the parent dir and env.build is the child dir.

Author: Am I looking in the wrong place when I see `build.env` in [`setupEnv`](https://github.com/buildpacks/lifecycle/blob/main/buildpack/build.go#L236)?

Author: I see it now as https://github.com/buildpacks/lifecycle/blob/main/platform/defaults.go#L85

Reviewer 2: > Am I looking in the wrong place when I see `build.env` in [`setupEnv`](https://github.com/buildpacks/lifecycle/blob/main/buildpack/build.go#L236)?

You are correct!

Reviewer 2: Ah well, now I'm confused... that function is for buildpack created env dirs. In my testing, `/cnb/build-config/env` worked.

| `<plan>`             | `CNB_PLAN_PATH`        | `<layers>/plan.toml`     | Path to resolved build plan (see [`plan.toml`](#plantoml-toml))                                 |
| `<platform>`         | `CNB_PLATFORM_DIR`     | `/platform`              | Path to platform directory                                                                      |
| `<uid>`              | `CNB_USER_ID`          |                          | UID of the build image `User`                                                                   |

> |                      | `CNB_BUILD_CONFIG_DIR` | `/cnb/config/env.build`  | Path to operator defined environment variables                                                  |

Reviewer 2: ```suggestion
| `<buildConfig>` | `CNB_BUILD_CONFIG_DIR` | `/cnb/build-config`                                | Path to operator-defined build config                                                                                                               |
```

> | `<build-config>`     | `CNB_BUILD_CONFIG_DIR` | `/cnb/build-config`      | Path to operator-defined environment variables                                                  |

Reviewer 2: ```suggestion
| `<build-config>`     | `CNB_BUILD_CONFIG_DIR` | `/cnb/build-config`      | Path to operator config directory                                                  |
```

Reviewer 2: This also needs to be added under `Usage`

/cnb/lifecycle/builder \
  [-app <app>] \
  [-buildpacks <buildpacks>] \

>   [-build-config <buildConfig>]

Reviewer: ```suggestion
  [-build-config <build-config>] \
```

| `<app>`          | `CNB_APP_DIR`          | `/workspace`          | Path to application directory                                                                  |
| `<buildpacks>`   | `CNB_BUILDPACKS_DIR`   | `/cnb/buildpacks`     | Path to buildpacks directory (see [Buildpacks Directory Layout](#buildpacks-directory-layout)) |
| `<group>`        | `CNB_GROUP_PATH`       | `<layers>/group.toml` | Path to group definition (see [`group.toml`](#grouptoml-toml))                                 |
| `<layers>`       | `CNB_LAYERS_DIR`       | `/layers`             | Path to layers directory                                                                       |
| `<log-level>`    | `CNB_LOG_LEVEL`        | `info`                | Log Level                                                                                      |
| `<plan>`         | `CNB_PLAN_PATH`        | `<layers>/plan.toml`  | Path to resolved build plan (see [`plan.toml`](#plantoml-toml))                                |
| `<platform>`     | `CNB_PLATFORM_DIR`     | `/platform`           | Path to platform directory                                                                     |

> | `<build-config>` | `CNB_BUILD_CONFIG_DIR` | `/cnb/build-config`   | Path to operator-defined environment variables                                                 |

Reviewer 2: ```suggestion
| `<build-config>` | `CNB_BUILD_CONFIG_DIR` | `/cnb/build-config`   | Path to operator config directory                                                 |
```

> | `<buildConfig>` | `CNB_BUILD_CONFIG_DIR` | `/cnb/config/env.build` | Path to operator defined environment variables                                                 |

Reviewer 2: ```suggestion
| `<buildConfig>` | `CNB_BUILD_CONFIG_DIR` | `/cnb/build-config`                                | Path to operator-defined build config                                                                                                               |
```

The platform SHOULD NOT set user-provided environment variables directly in the lifecycle execution environment.

The `<platform>/env/` directory follows the same convention as [Environment Variable Modification Rules](https://github.com/buildpacks/spec/blob/main/buildpack.md#environment-variable-modification-rules).

##### Operator-Defined Variables

> Operator-provided environment varaiables MUST be supplied by the platform as files in the `CNB_BUILD_CONFIG_DIR` directory.

Reviewer 2: ```suggestion
Operator-provided environment varaiables MUST be supplied by the platform as files in the `<build-config>/env/` directory.
```

Each file SHALL define a single environment variable, where the file name defines the key and the file contents define the value.

Operator-defined environment variables MAY be modified by subsequent buildpacks before they are provided to a given buildpack.

> The platform MUST set operator-provided environment variables directly in the lifecycle execution environment.

Reviewer 2: ```suggestion
The platform SHOULD NOT set operator-provided environment variables directly in the lifecycle execution environment.
```
Right?

> The `CNB_BUILD_CONFIG_DIR` directory follows the same convention as [Environment Variable Modification Rules](https://github.com/buildpacks/spec/blob/main/buildpack.md#environment-variable-modification-rules).

Reviewer: Can we add that if no extension is provided, the default extension is `.default`?

Reviewer 2: Added a suggestion

> The `CNB_BUILD_CONFIG_DIR` directory follows the same convention as [Environment Variable Modification Rules](https://github.com/buildpacks/spec/blob/main/buildpack.md#environment-variable-modification-rules).

Reviewer 2: ```suggestion
The `<build-config>/env/` directory follows the [Environment Variable Modification Rules](https://github.com/buildpacks/spec/blob/main/buildpack.md#environment-variable-modification-rules) outlined in the [Buildpack Interface Specification](buildpack.md), except for the modification behavior when no period-delimited suffix is provided; when no suffix is provided, the behavior is `default`.
```
